File: android/src/main/java/com/dodo/flutterbridge/data/DataCallNode.kt
package com.dodo.flutterbridge.data

import com.dodo.flutterbridge.call.CallNode
import com.dodo.flutterbridge.call.HandlerNode.StrategyData
import com.dodo.flutterbridge.call.strategy.HandlerStrategy
import com.dodo.flutterbridge.call.strategy.InvokerStrategy
import com.dodo.flutterbridge.call.strategy.SingleHandlerStrategy
import com.dodo.flutterbridge.call.strategy.SingleInvokerStrategy
import com.dodo.flutterbridge.common.Constant
import com.dodo.flutterbridge.model.FlutterCallInfo

/**
 *     author : liuduo
 *     e-mail : liuduo@gyenno.com
 *     time   : 2022/08/11
 *     desc   : data节点,处理type为data的逻辑
 *     version: 1.0
 */
object DataCallNode : CallNode<FlutterCallInfo, FlutterCallInfo> {

    override val name = Constant.Name.TYPE_DATA_NAME

    override val handlerStrategy: HandlerStrategy<FlutterCallInfo> =
        SingleHandlerStrategy(SingleHandlerStrategy.ConflictType.Exception)
    override val invokerStrategy: InvokerStrategy<FlutterCallInfo> =
        SingleInvokerStrategy(SingleInvokerStrategy.ConflictType.Exception)


    override fun decodeData(data: FlutterCallInfo): StrategyData<FlutterCallInfo> =
        StrategyData(
            data.methodInfo.name,
            true,
            data
        )


    override fun encodeData(data: FlutterCallInfo): FlutterCallInfo {
        data.methodInfo.type = name
        return data
    }
}

File: lib/src/data/data_call_node.dart
import 'package:flutter_bridge/src/call/call_leaf.dart';
import 'package:flutter_bridge/src/call/call_node.dart';
import 'package:flutter_bridge/src/call/call_root.dart';
import 'package:flutter_bridge/src/call/handler_node.dart';
import 'package:flutter_bridge/src/call/invoker_node.dart';
import 'package:flutter_bridge/src/call/strategy/handler_strategy.dart';
import 'package:flutter_bridge/src/call/strategy/invoker_strategy.dart';
import 'package:flutter_bridge/src/call/strategy/single_handler_strategy.dart';
import 'package:flutter_bridge/src/call/strategy/single_invoker_strategy.dart';
import 'package:flutter_bridge/src/common/constant.dart';
import 'package:flutter_bridge/src/model/flutter_call_info.dart';

///   author : liuduo
///   e-mail : liuduo@gyenno.com
///   time   : 2022/08/27
///   desc   : data节点,处理type为data的逻辑
///   version: 1.0
class DataCallNode
    with
        HandlerNode<FlutterCallInfo, FlutterCallInfo>,
        CallRoot<FlutterCallInfo, FlutterCallInfo>,
        InvokerNode<FlutterCallInfo, FlutterCallInfo>,
        CallLeaf<FlutterCallInfo, FlutterCallInfo>,
        CallNode<FlutterCallInfo, FlutterCallInfo> {
  @override
  HandlerStrategy<FlutterCallInfo> handlerStrategy =
      SingleHandlerStrategy(SingleHandlerConflictType.exception);

  @override
  InvokerStrategy<FlutterCallInfo> invokerStrategy =
      SingleInvokerStrategy(SingleInvokerConflictType.exception);

  @override
  String name = Constant.typeDataName;

  @override
  StrategyData<FlutterCallInfo> decodeData(FlutterCallInfo data) =>
      StrategyData(data.methodInfo.name, true, data);

  @override
  FlutterCallInfo encodeData(FlutterCallInfo data) {
    data.methodInfo.type = name;
    return data;
  }

  DataCallNode._();

  static final DataCallNode _instance = DataCallNode._();

  factory DataCallNode.instance() {
    return _instance;
  }
}
